sdk/keyvault/Azure.Security.KeyVault.Certificates/tests/LightweightPkcs8DecoderTests.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using Azure.Core;
+using NUnit.Framework;
+
+namespace Azure.Security.KeyVault.Certificates.Tests
+{
+    public class LightweightPkcs8DecoderTests
+    {
+        [Test]
+        public void VerifyECDecoderPrime256v1Imported() =>
+            // The following signature was generated in Key Vault using the certificate key corresponding to EcPrime256v1PrivateKeyImported,
+            // thus proving that Key Vault-generated OID 1.2.840.10045.1.1 (prime-field) is equivalent to OID 1.3.132.0.10 (secp256k1).
+            VerifyECDecoder(EcPrime256v1PrivateKeyImported, 256, @"Dqi/IwHjt2ttGyT0vleMvzsbStSOiDrevV4hhPGAXZoQ55b//keMmZ/weezuiSnliF4bTGWYxrqs73Yoj/6ddQ==");
+
+        [Test]
+        public void VerifyECDecoderSecp256k1() =>
+            VerifyECDecoder(EcSecp256k1PrivateKey, 256, @"YV+z2wZfDRqszLwcoPAnT4XmKMbfAqfwQYdYYU5FyicKWLtRmcaAX8Pd8h0OmSAdxoPO2sVyK2FP22pWq/6fJQ==");
+
+        [Test]
+        public void VerifyECDecoderPrime256v1() =>
+            VerifyECDecoder(EcPrime256v1PrivateKey, 256, @"1Byf+UuKVAEj+lsGBLQ4HiUDd4v/wa/XgcGlOEHorg47kjN2V6EDhC5aLwqZnXkbr48Y1umvBRAWcRYT6f8WfA==");
+
+        [Test]
+        public void VerifyECDecoderSecp384r1() =>
+            VerifyECDecoder(EcSecp384r1PrivateKey, 384, @"gHf3reGX71P7us9/fvcK1+6ONqRnkw5b9v+arz7G8iHk8BGflKwMlESIEEg2EezbNvLg6pURkCPC4OBngnM0OVYCjhb7Onh+zUogi2LSSCEIRZqrYcjzg8PHukp/sRyU");
+
+        [Test]
+        public void VerifyECDecoderSecp521r1() =>
+            VerifyECDecoder(EcSecp521r1PrivateKey, 521, @"ADB74QoZQcAbO0gfOhcavQLhV+nYMfLtNXC0i7a7FOYCiyZIJFoZtEzWbVTe4de1yMoe0lwGnW1mkCoPtEb0Q8f+AYAgRYSXlYkaF96XZYGcSp+RyALtOUpSvvpIaffk/3farwiFlGj6Dqryd5wLRksfLNR4DSyZbpEtHOg2RwkQ+vsI");
+
+        [Test]
+        public void VerifyECDecoderBadData()
+        {
+            byte[] data = Convert.FromBase64String(InvalidEcSecp521r1PrivateKey);
+
+            Assert.Throws<InvalidDataException>(() => LightweightPkcs8Decoder.DecodeECDsaPkcs8(data));
+        }
+
+        // Not using TestCaseSource because params too long for friendly test case rendering.
+        private static void VerifyECDecoder(string key, int keySize, string signature)
+        {
+#if NET461
+            Assert.Ignore("ECC is not supported before .NET Framework 4.7");
+#endif
+            byte[] data = Convert.FromBase64String(key);
+            byte[] signatureBytes = Convert.FromBase64String(signature);
+
+            using ECDsa ecdsa = LightweightPkcs8Decoder.DecodeECDsaPkcs8(data);
+
+            Assert.AreEqual(keySize, ecdsa.KeySize);
+            Assert.IsTrue(ecdsa.VerifyData(Encoding.UTF8.GetBytes("test"), signatureBytes, HashAlgorithmName.SHA256));
+        }
+
+        // This comes from Key Vault when a certificate using P-256K is generated
+        // with OID 1.2.840.10045.1.1 that is not a recognized curve.
```

Comment:
Please update every comment talking about that identifier to not call it a curve.

Reply:
Doesn't that OID identify the named curve, though? While, yes, I do need to handle the `prime-field` curve and parse the structure into an `ECCurve` for P-256K, I don't assume that Key Vault will always return that. For generated keys using P-256K, we do use a named OID of 1.3.132.0.10 because there is no additional `ECCurve` info stored in a JWK.

Reply:
> Doesn't that OID identify the named curve, though?

No, because you're in the "this isn't a named curve" path.  It identifies that the explicit parameters are associated with a prime curve, so the integer next to it represents the curve.P value, vs a Characteristic-2 curve's Polynomial value.

Reply:
The EC key structure is one of

* I'm a named curve, here's the curve identifier
* You should already know what curve I'm on, this value is null
* Instead of looking up the curve details by name, here are all the curve parameters:

```
SpecifiedECDomain(
    version, // what version of this structure is this payload
    fieldID.id, // what type of thing is fieldID.value
    fieldID.value,
    curve Curve (
        a,
        b,
        seed OPTIONAL),
    base,
    order,
    cofactor OPTIONAL,
    hash(AlgorithmIdentifier) OPTIONAL)
```

`1.2.840.10045.1.1 ` for fieldID.id just says that fieldID.value should be interpreted as a `Prime-p` value, which is just an alias for an `INTEGER` representing the prime P.

The other currently defined value, 1.2.840.10045.1.2, says the next value is a `Characteristic-two`, which is a more complicated structure describing the more complicated Char-2 field.

(What the structure here calls "Curve" is just "one curve in the field defined by the field formula", what everyone else calls "curve" includes the base point and field formula (which together also produce the order and cofactor))